tool_path_planner/src/halfedge_boundary_finder.cpp

```diff
@@ -0,0 +1,327 @@
+    {
+      boundary_he.clear();
+
+      circ = mesh.getInnerHalfEdgeAroundFaceCirculator(i);
+      circ_end = circ;
+      do
+      {
+        visited[pcl::geometry::toEdgeIndex(circ.getTargetIndex()).get()] = true;
+        boundary_he.push_back(circ.getTargetIndex());
+      } while (++circ != circ_end);
+
+      boundary_he_collection.push_back(boundary_he);
+    }
+  }
+}
+
+bool decimate(const pcl::PointCloud<pcl::PointNormal>& in, pcl::PointCloud<pcl::PointNormal>& out, double min_point_dist)
+{
+   out.clear();
+   out.push_back(in[0]);
+   using PType = std::remove_reference<decltype(in)>::type::PointType;
+   for(std::size_t i = 1 ; i < in.size() - 1; i++)
+   {
+     const PType& p = in[i];
+
+     double dist = (out.back().getVector3fMap() - p.getVector3fMap()).norm();
+     if(dist > min_point_dist)
+     {
+       out.push_back(p);
+     }
+   }
+   out.push_back(in.back());
+   return out.size() > 2;
+}
+
+void averageNormals(pcl::PointCloud<pcl::PointNormal>::ConstPtr src, pcl::PointCloud<pcl::PointXYZ>::ConstPtr src_xyz,
+                    pcl::PointCloud<pcl::PointNormal>& subset, double radius, double weight)
+{
+  using namespace pcl;
+  pcl::KdTreeFLANN<pcl::PointXYZ> kdtree;
+  kdtree.setInputCloud(src_xyz);
+  kdtree.setEpsilon(1e-3);
+  weight = std::abs(weight) > 1.0 ? 1.0 : std::abs(weight);
+  double r_2 = pow(radius, 2);
+  for(auto& p : subset)
+  {
+    std::vector<int> nearest_idx;
+    std::vector<float> nearest_dist_sqr;
+    PointXYZ pxyz;
+    pxyz.getVector3fMap() = p.getVector3fMap();
+    kdtree.radiusSearch(pxyz,radius,nearest_idx, nearest_dist_sqr);
+
+    // adding normals
+    std::size_t i = 0;
+    Eigen::Vector3f average_normal = std::accumulate(nearest_idx.begin(), nearest_idx.end(),p.getNormalVector3fMap(),
+                                                     [&i,
+                                                      &src,
+                                                      &nearest_dist_sqr,
+                                                      &weight,
+                                                      &r = radius,
+                                                      &r_2 ](Eigen::Vector3f n, int idx) -> Eigen::Vector3f{
+
+      if(idx > src->size())
+      {
+        return n;
+      }
+
+      double d_2 = nearest_dist_sqr[i++];
+      double f;
+      if(d_2 < 0.0 || d_2 > r_2)
+      {
+        f = 0.0;
+      }
+      else
+      {
+        f = (r - weight * std::sqrt( d_2 ))/ r;
+      }
+
+      Eigen::Vector3f v = (*src)[idx].getNormalVector3fMap();
+      v *= (f > 0 ? f : 0);
+
+      return n +  v;
+    });
+    p.getNormalVector3fMap() = average_normal.normalized();
+  }
+}
+
+namespace tool_path_planner
+{
+HalfEdgeBoundaryFinder::HalfEdgeBoundaryFinder()
+{
+
+}
+
+HalfEdgeBoundaryFinder::~HalfEdgeBoundaryFinder()
+{
+
+}
+
+void HalfEdgeBoundaryFinder::setInput(pcl::PolygonMesh::ConstPtr mesh)
```

Comment:
The other path generation methods did not set input, instead they passed the mesh as an argument. Which do you prefer and why?

Reply:
Good catch, I'll add the missing methods to make the interface consistent with the other class